pika/adapters/select_connection.py

```diff
@@ -264,20 +263,30 @@ def get_next_deadline(self):
         timeout = min((timeout, SelectPoller.POLL_TIMEOUT))
         return timeout * SelectPoller.POLL_TIMEOUT_MULT
 
+
     def process_timeouts(self):
         """Process the self._timeouts event stack"""
 
         now = time.time()
-        to_run = [timer for timer in self._timeouts.values()
-                  if timer['deadline'] <= now]
-
         # Run the timeouts in order of deadlines. Although this shouldn't
         # be strictly necessary it preserves old behaviour when timeouts
         # were only run periodically.
-        for t in sorted(to_run, key=itemgetter('deadline')):
-            t['callback']()
-            del self._timeouts[hash(frozenset(t.items()))]
-            self._next_timeout = None
+        to_run = sorted([(k, timer) for (k, timer) in self._timeouts.items()
```

Comment:
Thanks for pointing out sorted() can take a generator, which I was not aware of.
I think the code is better left as it is for now because it seems generator is not necessarily faster (especially because sorted() materialize generator as a list anyway?).
For our amusement, the script below generate following output. (The left side number is for generator and the right is for list.)

With threshold==0.01 (selectivity approx 1%)
Total = 0.4483 / 0.0092
Total = 0.3955 / 0.0099
With threshold==0.01 (selectivity approx 100%)
Total = 4.7233 / 2.2647
Total = 4.7220 / 2.2071

import time, random

def bench(l):
    start = time.time()
    sorted(l, key = lambda item: item[1]['k'])
    elapsed = time.time() - start
    return elapsed

def run(threshold):
    accum = {}
    for i in range(10):
        source = dict(map(lambda x: (random.random(), {'k':random.random()}),
                          range(0, 65536)))
        accum[1] = accum.get(1, 0.0) + \
            bench( (k,v) for (k,v) in source.items() if k<threshold )
        accum[2] = accum.get(2, 0.0) + \
            bench( [(k,v) for (k,v) in source.items() if k<threshold] )
    print 'Total = %.4f / %.4f' % (accum[1], accum[2])

    accum = {}
    for i in range(10):
        source = dict(map(lambda x: (random.random(), {'k':random.random()}),
                          range(0, 65536)))
        accum[2] = accum.get(2, 0.0) + \
            bench( [(k,v) for (k,v) in source.items() if k<threshold] )
        accum[1] = accum.get(1, 0.0) + \
            bench( (k,v) for (k,v) in source.items() if k<threshold )
    print 'Total = %.4f / %.4f' % (accum[1], accum[2])

if __name__=='__main__':
    print 'With threshold==0.01 (selectivity approx 1%)'
    run(0.01)
    print 'With threshold==0.01 (selectivity approx 100%)'
    run(1.0)